Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,726 @@
+{
+    auto result = adopt_ref_if_nonnull(new (nothrow) ISO9660Inode(fs, record, name));
+    if (!result) {
+        return ENOMEM;
+    }
+    return result.release_nonnull();
+}
+
+void ISO9660Inode::create_metadata()
+{
+    u32 data_length = LittleEndian { m_record.data_length.little };
+    bool is_directory = has_flag(m_record.file_flags, ISO::FileFlags::Directory);
+    time_t recorded_at = parse_numerical_date_time(m_record.recording_date_and_time);
+
+    m_metadata = {
+        .inode = identifier(),
+        .size = data_length,
+        .mode = static_cast<mode_t>((is_directory ? S_IFDIR : S_IFREG) | (is_directory ? 0555 : 0444)),
+        .uid = 0,
+        .gid = 0,
+        .link_count = 1,
+        .atime = recorded_at,
+        .ctime = recorded_at,
+        .mtime = recorded_at,
+        .dtime = 0,
+        .block_count = 0,
+        .block_size = 0,
+        .major_device = 0,
+        .minor_device = 0,
+    };
+}
+
+time_t ISO9660Inode::parse_numerical_date_time(ISO::NumericalDateAndTime const& date)
+{
+    i32 year_offset = date.years_since_1900 - 70;
+
+    return (year_offset * 60 * 60 * 24 * 30 * 12)
+        + (date.month * 60 * 60 * 24 * 30)
+        + (date.day * 60 * 60 * 24)
+        + (date.hour * 60 * 60)
+        + (date.minute * 60)
+        + date.second;
+}
+
+StringView ISO9660Inode::get_normalized_filename(ISO::DirectoryRecordHeader const* record, Bytes buffer)
+{
+    auto file_identifier = reinterpret_cast<u8 const*>(record + 1);
+    auto filename = StringView { file_identifier, record->file_identifier_length };
+
+    if (filename.length() == 1) {
+        if (filename[0] == '\0') {
+            filename = "."sv;
+        }
+
+        if (filename[0] == '\1') {
+            filename = ".."sv;
+        }
+    }
+
+    if (!has_flag(record->file_flags, ISO::FileFlags::Directory)) {
+        // FIXME: We currently strip the file version from the filename,
+        //        but that may be used later down the line if the file actually
+        //        has mulitple versions on the disk.
+        Optional<size_t> semicolon = filename.find(';');
+        if (semicolon.has_value()) {
+            filename = filename.substring_view(0, semicolon.value());
+        }
+
+        if (filename[filename.length() - 1] == '.') {
+            filename = filename.substring_view(0, filename.length() - 1);
+        }
+    }
+
+    if (filename.length() > buffer.size()) {
+        // FIXME: Rock Ridge allows filenames up to 255 characters, so we should
+        //        probably support that instead of truncating.
+        filename = filename.substring_view(0, buffer.size());
+    }
+
+    for (size_t i = 0; i < filename.length(); i++) {
+        buffer[i] = to_ascii_lowercase(filename[i]);
+    }
+
+    return { buffer.data(), filename.length() };
+}
+
+InodeIndex ISO9660Inode::get_inode_index(ISO::DirectoryRecordHeader const* record, StringView const& name)
```

Comment:
This function does not support `record` being null, so it should take `record` as a reference.